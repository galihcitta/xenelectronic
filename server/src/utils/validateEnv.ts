import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
