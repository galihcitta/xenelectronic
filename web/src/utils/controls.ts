import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
