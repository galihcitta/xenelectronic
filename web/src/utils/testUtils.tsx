import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
