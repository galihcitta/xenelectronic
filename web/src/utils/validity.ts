import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
