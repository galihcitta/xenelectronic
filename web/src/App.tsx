import React, { lazy, Suspense } from 'react';

import { GlobalStyles } from './utils/styles';
import { GlobalSpinner } from './components/GlobalSpinner/GlobalSpinner.component';
import { withSession } from './hoc/withSession';

const AuthApp = lazy(() => import('./AuthApp'));
const UnauthApp = lazy(() => import('./UnauthApp'));

export interface AppProps {
  isAuth: boolean | null;
}

export const _App: React.FC<AppProps> = ({ isAuth }) => (
  <>
    <GlobalStyles />
    <Suspense fallback={<GlobalSpinner />}>
      {isAuth === null && <GlobalSpinner />}
      {isAuth === false && <UnauthApp />}
      {isAuth === true && <AuthApp />}
    </Suspense>
  </>
);

export const App = withSession(_App);
