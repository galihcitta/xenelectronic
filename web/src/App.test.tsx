import { render, wait, waitForElement } from '@testing-library/react';
import { _App, AppProps } from './App';
import { withProvider } from './utils/testUtils';

const AppWithProvider = withProvider(_App);

describe('<App />', () => {
  const props: AppProps = {
    isAuth: false
  };

  const setup = (props: AppProps) => {
    const utils = render(AppWithProvider(props));
    return { ...utils };
  };

  it('renders <App /> component', async () => {
    const { asFragment } = setup(props);
    await wait(); // wait for import to be resolved

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders <Spinner /> when `isAuth` prop is null', async () => {
    const { getByTestId } = setup({ ...props, isAuth: null });
    const spinner = await waitForElement(() => getByTestId('Spinner'));

    expect(spinner).toBeInTheDocument();
  });

  it('renders <AuthApp /> when `isAuth` prop is true', async () => {
    const { getByText } = setup({ ...props, isAuth: true });
    await wait();
    const homeLink = getByText(/home/i).closest('a');

    expect(homeLink).toHaveAttribute('href', '/');
  });
});
