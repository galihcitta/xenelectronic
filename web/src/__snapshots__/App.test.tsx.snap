// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders <App /> component 1`] = `
<DocumentFragment>
  <div
    class="sc-htpNat fjcJqy"
  >
    <div
      class="sc-kGXeez kGgkhh"
      data-testid="Overlay"
    />
    <nav
      class="sc-hMqMXs eiWHUL"
      data-testid="Wrapper"
    >
      <header
        class="sc-kEYyzF csDjny"
      >
        <p
          class="sc-kkGfuU BCNoT"
        >
          Interior & Exterior
        </p>
        <button
          aria-label="close"
          class="sc-iAyFgw fsZTAk"
          title="close menu"
        >
          <i
            class="far fa-times-circle"
          />
        </button>
      </header>
      <ul
        class="sc-hSdWYo fMsZYH"
      >
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-ckVGcZ dgJLfB"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <button
            class="sc-dxgOiQ cMyJyP"
          >
            Interior
            <i
              class="fas fa-angle-right"
            />
          </button>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <button
            class="sc-dxgOiQ cMyJyP"
          >
            Exterior
            <i
              class="fas fa-angle-right"
            />
          </button>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-ckVGcZ dgJLfB"
            href="/account"
          >
            Account
          </a>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-ckVGcZ dgJLfB"
            href="/cart"
          >
            Cart (0)
          </a>
        </li>
      </ul>
    </nav>
    <nav
      class="sc-hMqMXs kHxeFj"
      data-testid="Wrapper"
    >
      <header
        class="sc-kEYyzF csDjny"
      >
        <p
          class="sc-kkGfuU BCNoT"
        >
          Interior
        </p>
        <button
          aria-label="close"
          class="sc-iAyFgw fsZTAk"
          title="close menu"
        >
          <i
            class="far fa-times-circle"
          />
        </button>
      </header>
      <ul
        class="sc-hSdWYo fMsZYH"
      >
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-eNQAEJ dqipbD"
            href="/shop"
          >
            Chairs
          </a>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-eNQAEJ dqipbD"
            href="/shop"
          >
            Lamps
          </a>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-eNQAEJ dqipbD"
            href="/shop"
          >
            Clocks
          </a>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-eNQAEJ dqipbD"
            href="/shop"
          >
            Vases
          </a>
        </li>
      </ul>
    </nav>
    <nav
      class="sc-hMqMXs kHxeFj"
      data-testid="Wrapper"
    >
      <header
        class="sc-kEYyzF csDjny"
      >
        <p
          class="sc-kkGfuU BCNoT"
        >
          Exterior
        </p>
        <button
          aria-label="close"
          class="sc-iAyFgw fsZTAk"
          title="close menu"
        >
          <i
            class="far fa-times-circle"
          />
        </button>
      </header>
      <ul
        class="sc-hSdWYo fMsZYH"
      >
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-eNQAEJ dqipbD"
            href="/shop"
          >
            Benches
          </a>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-eNQAEJ dqipbD"
            href="/shop"
          >
            Flowerpots
          </a>
        </li>
        <li
          class="sc-kpOJdX eKfpAE"
        >
          <a
            class="sc-eNQAEJ dqipbD"
            href="/shop"
          >
            Plants
          </a>
        </li>
      </ul>
    </nav>
    <header
      class="sc-gqjmRU eeYKtP"
    >
      <div
        class="sc-VigVT dgUaaW"
      >
        <a
          class="sc-jTzLTM eFukZp"
          href="/"
        >
          <h1
            class="sc-fjdhpX gVNPhm"
          >
            Interior & Exterior
          </h1>
        </a>
        <nav
          class="sc-jzJRlG cbWyOs"
        >
          <button
            aria-label="open menu"
            class="sc-cSHVUG buEebX"
            title="Menu"
          >
            <i
              class="fas fa-bars"
            />
          </button>
          <ul
            class="sc-ifAKCX dqpNVd"
          >
            <li
              class="sc-gzVnrw dRxXhy"
            >
              <a
                class="sc-bZQynM hKSVIJ"
                href="/shop"
              >
                Products
                <i
                  class="fas fa-caret-down"
                />
              </a>
              <div
                class="sc-EHOje hJLuSz"
              >
                <ul
                  class="sc-htoDjs kKCJWi"
                >
                  <li
                    class="sc-dnqmqq gjUOyw"
                  >
                    <a
                      class="sc-iwsKbI iQdAId"
                      href="/shop"
                    >
                      Chairs
                    </a>
                  </li>
                  <li
                    class="sc-dnqmqq gjUOyw"
                  >
                    <a
                      class="sc-iwsKbI iQdAId"
                      href="/shop"
                    >
                      Lamps
                    </a>
                  </li>
                  <li
                    class="sc-dnqmqq gjUOyw"
                  >
                    <a
                      class="sc-iwsKbI iQdAId"
                      href="/shop"
                    >
                      Clocks
                    </a>
                  </li>
                  <li
                    class="sc-dnqmqq gjUOyw"
                  >
                    <a
                      class="sc-iwsKbI iQdAId"
                      href="/shop"
                    >
                      Vases
                    </a>
                  </li>
                  <li
                    class="sc-dnqmqq gjUOyw"
                  >
                    <a
                      class="sc-iwsKbI iQdAId"
                      href="/shop"
                    >
                      Benches
                    </a>
                  </li>
                  <li
                    class="sc-dnqmqq gjUOyw"
                  >
                    <a
                      class="sc-iwsKbI iQdAId"
                      href="/shop"
                    >
                      Flowerpots
                    </a>
                  </li>
                  <li
                    class="sc-dnqmqq gjUOyw"
                  >
                    <a
                      class="sc-iwsKbI iQdAId"
                      href="/shop"
                    >
                      Plants
                    </a>
                  </li>
                </ul>
              </div>
            </li>
            <li
              class="sc-gzVnrw dRxXhy"
            >
              <a
                class="sc-bZQynM hKSVIJ"
                href="/account"
              >
                Account
              </a>
            </li>
            <li
              class="sc-gzVnrw dRxXhy"
            >
              <a
                class="sc-bZQynM hKSVIJ"
                href="/cart"
              >
                Cart (0)
              </a>
            </li>
          </ul>
          <button
            aria-label="submit searching"
            class="sc-kAzzGY bmJMdz"
            title="Search"
          >
            <i
              class="fas fa-search"
            />
          </button>
        </nav>
      </div>
    </header>
    <form
      class="sc-eHgmQL duFWef"
    >
      <input
        class="sc-cvbbAY dBOhuc"
        placeholder="Search for products ..."
        type="test"
        value=""
      />
      <button
        class="sc-jWBwVP eCsuXW"
        title="submit"
        type="submit"
      >
        <i
          class="far fa-arrow-alt-circle-right"
        />
      </button>
    </form>
    <main
      class="sc-bxivhb hKOuQv"
    >
      <section
        class="sc-brqgnP kiahNM"
      >
        <a
          class="sc-jDwBTQ lnSPoE"
          href="/shop"
        >
          <h2
            class="sc-cMljjf jsdkcN"
          >
            Lamps
          </h2>
          <div
            class="sc-jAaTju cEjMrU"
          >
            <img
              alt=""
              class="sc-gPEVay jRHnne"
              src="lamp.png"
            />
          </div>
        </a>
        <a
          class="sc-jDwBTQ lnSPoE"
          href="/shop"
        >
          <h2
            class="sc-cMljjf jsdkcN"
          >
            Vases
          </h2>
          <div
            class="sc-jAaTju cEjMrU"
          >
            <img
              alt=""
              class="sc-gPEVay jRHnne"
              src="vase.png"
            />
          </div>
        </a>
        <a
          class="sc-jDwBTQ lnSPoE"
          href="/shop"
        >
          <h2
            class="sc-cMljjf jsdkcN"
          >
            Chairs
          </h2>
          <div
            class="sc-jAaTju cEjMrU"
          >
            <img
              alt=""
              class="sc-gPEVay jRHnne"
              src="chair.png"
            />
          </div>
        </a>
        <a
          class="sc-jDwBTQ lnSPoE"
          href="/shop"
        >
          <h2
            class="sc-cMljjf jsdkcN"
          >
            Clocks
          </h2>
          <div
            class="sc-jAaTju cEjMrU"
          >
            <img
              alt=""
              class="sc-gPEVay jRHnne"
              src="clock.png"
            />
          </div>
        </a>
        <a
          class="sc-jDwBTQ lnSPoE"
          href="/shop"
        >
          <h2
            class="sc-cMljjf jsdkcN"
          >
            Benches
          </h2>
          <div
            class="sc-jAaTju cEjMrU"
          >
            <img
              alt=""
              class="sc-gPEVay jRHnne"
              src="bench.png"
            />
          </div>
        </a>
        <a
          class="sc-jDwBTQ lnSPoE"
          href="/shop"
        >
          <h2
            class="sc-cMljjf jsdkcN"
          >
            Flowerpots
          </h2>
          <div
            class="sc-jAaTju cEjMrU"
          >
            <img
              alt=""
              class="sc-gPEVay jRHnne"
              src="flowerpot.png"
            />
          </div>
        </a>
        <a
          class="sc-jDwBTQ lnSPoE"
          href="/shop"
        >
          <h2
            class="sc-cMljjf jsdkcN"
          >
            Plants
          </h2>
          <div
            class="sc-jAaTju cEjMrU"
          >
            <img
              alt=""
              class="sc-gPEVay jRHnne"
              src="plant.png"
            />
          </div>
        </a>
      </section>
    </main>
    <footer
      class="sc-chPdSV fDjWgM"
    >
      <p
        class="sc-kgoBCf gATMll"
      >
        - Interior & Exterior -
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
